controllers/fenceagentsremediation_controller.go

```diff
@@ -68,69 +75,68 @@ type FenceAgentsRemediationReconciler struct {
 			// FAR is deleted, stop reconciling
-			r.Log.Info("Fence Agents Remediation not found, nothing to do")
-			return ctrl.Result{}, nil
+			r.Log.Info("FAR CR is deleted - nothing to do", "CR Name", req.Name, "CR Namespace", req.Namespace)
+			return emptyResult, nil
 		}
-		r.Log.Error(err, "failed to get FAR")
-		return ctrl.Result{}, err
+		r.Log.Error(err, "failed to get FAR CR")
+		return emptyResult, err
 	}
-	key := client.ObjectKey{Namespace: req.Namespace, Name: fenceAgentsTemplateName}
-	farTemplate := &v1alpha1.FenceAgentsRemediationTemplate{}
-	if err := r.Get(ctx, key, farTemplate); err != nil {
-		r.Log.Error(err, "failed to get FAR template")
-		return ctrl.Result{}, err
+	// TODO: Validate FAR CR name to nodeName. Run isNodeNameValid
+	// Fetch the FAR's pod
+	r.Log.Info("Fetch FAR's pod")
+	pod, err := r.getFenceAgentsPod(req.Namespace)
+	if err != nil {
+		return emptyResult, err
 	}
 
-	pod, err := r.getFAPod(req.NamespacedName.Namespace)
+	// Build CLI executer for FAR's pod
+	r.Log.Info("Build CLI executer for FAR's pod")
+	ex, err := r.ExecutorHook(pod)
 	if err != nil {
-		return ctrl.Result{}, err
+		return emptyResult, err
 	}
 
-	ex, err := cli.NewExecuter(pod)
+	//TODO: Check that FA is excutable? run cli.IsExecuteable
+
+	r.Log.Info("Create and execute the fence agent", "Fence Agent", farTemplate.Spec.Agent)
+	faParams, err := buildFenceAgentParams(farTemplate, far)
 	if err != nil {
-		return ctrl.Result{}, err
+		return emptyResult, err
 	}
-
-	faParams := buildFenceAgentParams(farTemplate, far)
 	cmd := append([]string{farTemplate.Spec.Agent}, faParams...)
-	//fence_ipmilan --ip=192.168.111.1 --ipport=6233 --username=admin --password=password --action=status --lanplus --verbose
+	// The Fence Agent is excutable and the parameters are valid but we don't know about their values
 	if _, _, err := ex.Execute(cmd); err != nil {
-		return ctrl.Result{}, err
-	}
-
-	return ctrl.Result{}, nil
-}
-
-func buildFenceAgentParams(farTemplate *v1alpha1.FenceAgentsRemediationTemplate, far *v1alpha1.FenceAgentsRemediation) []string {
-	var fenceAgentParams []string
-	for paramName, paramVal := range farTemplate.Spec.SharedParameters {
-		fenceAgentParams = appendParamToSlice(fenceAgentParams, string(paramName), paramVal)
-
+		//TODO: better seperation between errors from wrong shared parameters values and wrong node parameters values
+		return emptyResult, err
 	}
-
-	nodeName := v1alpha1.NodeName(far.Name)
-	for paramName, nodeMap := range farTemplate.Spec.NodeParameters {
-		fenceAgentParams = appendParamToSlice(fenceAgentParams, string(paramName), nodeMap[nodeName])
-	}
-
-	return fenceAgentParams
-}
-
-// SetupWithManager sets up the controller with the Manager.
-func (r *FenceAgentsRemediationReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1alpha1.FenceAgentsRemediation{}).
-		Complete(r)
+	r.Log.Info("Finish FenceAgentsRemediation Reconcile")
```

Comment:
I suggest putting this in `defer` at the start otherwise it'll not be printed in case of an early exit.

Reply:
Oho, now I understand the advantage of `defer`.